staging/src/k8s.io/apiserver/pkg/endpoints/filters/constrained_impersonation.go

```diff
@@ -0,0 +1,292 @@
+			// fallback to use legacy impersonation
+			legacyDecision, legacyReason, err := authorizeLegacyImpersonation(req, actingAsAttrsList, a)
+			if err != nil || legacyDecision != authorizer.DecisionAllow {
+				klog.V(4).InfoS("Forbidden", "URI", req.RequestURI, "reason", legacyReason, "err", err)
+				responsewriters.Forbidden(ctx, actingAsAttributes, w, req, reason, s)
+				return
+			}
+		}
+
+		if username != user.Anonymous {
+			// When impersonating a non-anonymous user, include the 'system:authenticated' group
+			// in the impersonated user info:
+			// - if no groups were specified
+			// - if a group has been specified other than 'system:authenticated'
+			//
+			// If 'system:unauthenticated' group has been specified we should not include
+			// the 'system:authenticated' group.
+			addAuthenticated := true
+			for _, group := range groups {
+				if group == user.AllAuthenticated || group == user.AllUnauthenticated {
+					addAuthenticated = false
+					break
+				}
+			}
+
+			if addAuthenticated {
+				groups = append(groups, user.AllAuthenticated)
+			}
+		} else {
+			addUnauthenticated := true
+			for _, group := range groups {
+				if group == user.AllUnauthenticated {
+					addUnauthenticated = false
+					break
+				}
+			}
+
+			if addUnauthenticated {
+				groups = append(groups, user.AllUnauthenticated)
+			}
+		}
+
+		newUser := &user.DefaultInfo{
+			Name:   username,
+			Groups: groups,
+			Extra:  userExtra,
+			UID:    uid,
+		}
+		req = req.WithContext(request.WithUser(ctx, newUser))
+
+		oldUser, _ := request.UserFrom(ctx)
+		httplog.LogOf(req, w).Addf("%v is impersonating %v", userString(oldUser), userString(newUser))
+
+		audit.LogImpersonatedUser(audit.WithAuditContext(ctx), newUser)
+
+		// clear all the impersonation headers from the request
+		req.Header.Del(authenticationv1.ImpersonateUserHeader)
+		req.Header.Del(authenticationv1.ImpersonateGroupHeader)
+		req.Header.Del(authenticationv1.ImpersonateUIDHeader)
+		for headerName := range req.Header {
+			if strings.HasPrefix(headerName, authenticationv1.ImpersonateUserExtraHeaderPrefix) {
+				req.Header.Del(headerName)
+			}
+		}
+
+		handler.ServeHTTP(w, req)
+	})
+}
+
+func authorizeConstrainedImpersonation(ctx context.Context, requestorAttrs *authorizer.AttributesRecord, targetAttributesList []*authorizer.AttributesRecord, a authorizer.Authorizer) (authorizer.Decision, string, authorizer.Attributes, error) {
+	for _, actingAsAttributes := range targetAttributesList {
+		// group and version are always authentication/v1 with constrained impersonation.
+		actingAsAttributesCopy := *actingAsAttributes
+		actingAsAttributesCopy.APIGroup = authenticationv1.SchemeGroupVersion.Group
+		actingAsAttributesCopy.APIVersion = authenticationv1.SchemeGroupVersion.Version
+
+		// authorize node verb if the impersonated user is node
+		if actingAsAttributes.Resource == "users" && strings.HasPrefix(actingAsAttributesCopy.Name, "system:node:") {
```

Comment:
I would expect `nodes` to be a first class concept just like users and SAs.

Reply:
updated with some refactor.